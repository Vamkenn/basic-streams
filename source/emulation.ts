import {Stream} from "./index"

export class Value<T> {
  constructor(
    public readonly value: T,
    public readonly cb?: ((x: T) => void),
  ) {}

  callCb() {
    if (this.cb) {
      this.cb(this.value)
    }
  }

  toJSON() {
    return this.value
  }
}

export class TimeSpan {
  constructor(public readonly ms: number) {}

  toJSON() {
    return `after ${this.ms} ms...`
  }
}

type TimelineItem<T> = TimeSpan | Value<T>

export class Timeline<T> {
  constructor(
    public readonly items: ReadonlyArray<TimelineItem<T>>,
    dontCompact?: boolean,
  ) {
    if (dontCompact) {
      return this
    }

    const finalItems: Array<TimelineItem<T>> = []
    for (const item of this.items) {
      if (item instanceof TimeSpan) {
        if (item.ms === 0) {
          continue
        }
        if (finalItems.length > 0) {
          const last = finalItems[finalItems.length - 1]
          if (last instanceof TimeSpan) {
            finalItems[finalItems.length - 1] = new TimeSpan(last.ms + item.ms)
            continue
          }
        }
      }
      finalItems.push(item)
    }

    if (
      finalItems.length > 0 &&
      finalItems[finalItems.length - 1] instanceof TimeSpan
    ) {
      finalItems.pop()
    }

    return new Timeline(finalItems, true)
  }

  withCb(cb: (value: T) => void): Timeline<T> {
    return new Timeline(
      this.items.map(
        item => (item instanceof TimeSpan ? item : new Value(item.value, cb)),
      ),
    )
  }

  merge<U>(
    another: Timeline<U>,
    result: Timeline<T | U> = new Timeline([]),
  ): Timeline<T | U> {
    const a = this
    const b = another
    if (a.items.length === 0 && b.items.length === 0) {
      return result
    }
    if (a.items.length === 0) {
      return new Timeline([
        ...result.items,
        ...(b.items as Array<TimelineItem<T | U>>),
      ])
    }
    if (b.items.length === 0) {
      return new Timeline([
        ...result.items,
        ...(a.items as Array<TimelineItem<T | U>>),
      ])
    }
    const timeToValueA = a.timeToValue()
    const timeToValueB = b.timeToValue()
    if (timeToValueA === Infinity && timeToValueB === Infinity) {
      return new Timeline([
        new TimeSpan(Math.max(a.timeToEnd(), b.timeToEnd())),
        ...result.items,
      ])
    }
    const timeToValue = Math.min(timeToValueA, timeToValueB)
    const subtractA = a.subtractTime(timeToValue)
    const subtractB = b.subtractTime(timeToValue)
    return subtractA.rest.merge(
      subtractB.rest,
      new Timeline([
        ...result.items,
        new TimeSpan(timeToValue),
        ...(subtractA.values as Array<Value<T | U>>),
        ...(subtractB.values as Array<Value<T | U>>),
      ]),
    )
  }

  isEmpty(): boolean {
    return this.items.length === 0
  }

  takeOne(): {item: TimelineItem<T>; rest: Timeline<T>} {
    if (this.isEmpty()) {
      throw new Error("cannot takeOne() from an empty schedule")
    }
    const [top, ...rest] = this.items
    return {item: top, rest: new Timeline(rest)}
  }

  subtractTime(ms: number): {values: Value<T>[]; rest: Timeline<T>} {
    if (this.isEmpty()) {
      return {rest: this, values: []}
    }
    const {item, rest} = this.takeOne()
    if (item instanceof Value) {
      const r = rest.subtractTime(ms)
      return {rest: r.rest, values: [item, ...r.values]}
    }
    return item.ms <= ms
      ? rest.subtractTime(ms - item.ms)
      : {
          rest: new Timeline([new TimeSpan(item.ms - ms), ...rest.items]),
          values: [],
        }
  }

  timeToEnd(): number {
    return this.items.reduce(
      (result, x) => result + (x instanceof TimeSpan ? x.ms : 0),
      0,
    )
  }

  timeToValue(): number {
    if (this.isEmpty()) {
      return Infinity
    }
    const {item, rest} = this.takeOne()
    return item instanceof Value ? 0 : item.ms + rest.timeToValue()
  }

  toJSON() {
    return this.items
  }
}

type EmulationGenerator<T> = (
  createStream: <U>(...scheduleItems: Array<TimelineItem<U>>) => Stream<U>,
) => Stream<T>

export function t(ms: number): TimeSpan {
  return new TimeSpan(ms)
}

export function v<T>(x: T): Value<T> {
  return new Value(x)
}

export function emulate<T>(generator: EmulationGenerator<T>): Timeline<T> {
  const state = {
    schedule: new Timeline([]),
    result: [] as Array<TimelineItem<T>>,
  }

  const resultStream = generator((...scheduleItems) => {
    return cb => {
      state.schedule = state.schedule.merge(
        new Timeline(scheduleItems).withCb(cb),
      )
      return () => {
        state.schedule = new Timeline(
          state.schedule.items.filter(
            item => item instanceof TimeSpan || item.cb !== cb,
          ),
        )
      }
    }
  })

  resultStream(value => state.result.push(new Value(value)))

  while (state.schedule.timeToValue() !== Infinity) {
    const time = state.schedule.timeToValue()
    const {rest, values} = state.schedule.subtractTime(time)
    state.schedule = rest
    state.result.push(new TimeSpan(time))
    values.forEach(item => item.callCb())
  }

  return new Timeline(state.result)
}
