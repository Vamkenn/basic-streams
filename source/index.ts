export type Stream<T> = (cb: (payload: T) => void) => (() => void)
export type StreamLoose<T> = (
  cb: (payload: T, ...rest: any[]) => any,
  ...rest: any[]
) => any

type Maybe<T> = {tag: "nothing"} | {tag: "just"; value: T}

function noop() {}

function id<T>(x: T): T {
  return x
}

export function empty(): Stream<any> {
  return emptyStream
}
const emptyStream = () => noop

export function of<T>(x: T): Stream<T> {
  return cb => {
    cb(x)
    return noop
  }
}

export function fromIterable<T>(xs: Iterable<T>): Stream<T> {
  return cb => {
    for (const x of xs) {
      cb(x)
    }
    return noop
  }
}

export function fromLoose<T>(looseStream: StreamLoose<T>): Stream<T> {
  return _cb => {
    let cb: null | ((x: T) => any) = _cb
    let disposer = looseStream(x => {
      if (cb !== null) {
        cb(x)
      }
    })
    return () => {
      if (cb === null) {
        return
      }
      cb = null
      if (typeof disposer === "function") {
        disposer()
      }
      disposer = null
    }
  }
}

export function map<T, U>(fn: (x: T) => U, stream: Stream<T>): Stream<U> {
  return cb => stream(payload => cb(fn(payload)))
}

export function filter<T>(
  predicate: (x: T) => boolean,
  stream: Stream<T>,
): Stream<T> {
  return cb =>
    stream(payload => {
      if (predicate(payload)) cb(payload)
    })
}

export function chain<T, U>(
  fn: (x: T) => Stream<U>,
  stream: Stream<T>,
): Stream<U> {
  return cb => {
    let spawnedDisposers: Array<() => void> = []
    const mainDisposer = stream(payload => {
      spawnedDisposers.push(fn(payload)(cb))
    })
    return () => {
      spawnedDisposers.forEach(fn => fn())
      mainDisposer()
    }
  }
}

export function chainLatest<T, U>(
  fn: (x: T) => Stream<U>,
  stream: Stream<T>,
): Stream<U> {
  return cb => {
    let spawnedDisposer = () => {}
    const mainDisposer = stream(payload => {
      spawnedDisposer()
      spawnedDisposer = fn(payload)(cb)
    })
    return () => {
      spawnedDisposer()
      mainDisposer()
    }
  }
}

export function ap<T, U>(
  streamF: Stream<(x: T) => U>,
  streamV: Stream<T>,
): Stream<U> {
  return cb => {
    let latestF: Maybe<(x: T) => U> = {tag: "nothing"}
    let latestV: Maybe<T> = {tag: "nothing"}
    const push = () => {
      if (latestF.tag === "just" && latestV.tag === "just") {
        const fn = latestF.value
        cb(fn(latestV.value))
      }
    }
    const disposef = streamF(f => {
      latestF = {tag: "just", value: f}
      push()
    })
    const disposev = streamV(v => {
      latestV = {tag: "just", value: v}
      push()
    })
    return () => {
      disposef()
      disposev()
    }
  }
}

export function map2<A, B, C>(
  fn: (a: A, b: B) => C,
  streamA: Stream<A>,
  streamB: Stream<B>,
): Stream<C> {
  return ap(map((a: A) => (b: B) => fn(a, b), streamA), streamB)
}

export function map3<A, B, C, D>(
  fn: (a: A, b: B, c: C) => D,
  streamA: Stream<A>,
  streamB: Stream<B>,
  streamC: Stream<C>,
): Stream<D> {
  return ap(
    ap(map((a: A) => (b: B) => (c: C) => fn(a, b, c), streamA), streamB),
    streamC,
  )
}

export function combineArray<T>(streams: Array<Stream<T>>): Stream<Array<T>> {
  return streams.reduce(liftedAppend, of([]))
}
function liftedAppend<T>(xs: Stream<T[]>, x: Stream<T>): Stream<T[]> {
  return map2<T[], T, T[]>(append, xs, x)
}
function append<T>(xs: T[], x: T): T[] {
  return xs.concat([x])
}

export function merge<T>(streams: Array<Stream<T>>): Stream<T> {
  return chain<Stream<T>, T>(id, fromIterable(streams))
}

export function scan<N, A>(
  reducer: (acc: A, next: N) => A,
  seed: A,
  stream: Stream<N>,
): Stream<A> {
  return cb => {
    let acc = seed
    cb(acc)
    return stream(next => {
      acc = reducer(acc, next)
      cb(acc)
    })
  }
}

export function take<T>(n: number, stream: Stream<T>): Stream<T> {
  return cb => {
    let count = 0
    let disposer: (() => void) | null = null
    const dispose = () => {
      if (disposer !== null) {
        disposer()
        disposer = null
      }
    }
    disposer = stream(x => {
      count++
      if (count <= n) {
        cb(x)
      }
      if (count >= n) {
        dispose()
      }
    })
    if (count >= n) {
      dispose()
    }
    return dispose
  }
}

export function takeWhile<T>(
  predicate: (x: T) => boolean,
  stream: Stream<T>,
): Stream<T> {
  return cb => {
    let completed = false
    let disposer: (() => void) | null = null
    const dispose = () => {
      if (disposer !== null) {
        disposer()
        disposer = null
      }
    }
    disposer = stream(x => {
      if (!completed) {
        if (predicate(x)) {
          cb(x)
        } else {
          completed = true
          dispose()
        }
      }
    })
    if (completed) {
      dispose()
    }
    return dispose
  }
}

export function takeUntil<T>(
  controller: Stream<any>,
  stream: Stream<T>,
): Stream<T> {
  return cb => {
    let disposed = false
    let mainDisposer: (() => void) | null = null
    let ctrlDisposer: (() => void) | null = null
    const dispose = () => {
      if (disposed) {
        return
      }
      disposed = true
      if (mainDisposer !== null) {
        mainDisposer()
        mainDisposer = null
      }
      if (ctrlDisposer !== null) {
        ctrlDisposer()
        ctrlDisposer = null
      }
    }

    ctrlDisposer = controller(dispose)

    if (disposed && ctrlDisposer !== null) {
      ctrlDisposer()
    }

    if (disposed) {
      // subscribe anyway for consistency
      stream(noop)()
    } else {
      mainDisposer = stream(cb)
    }

    return dispose
  }
}

export function skip<T>(n: number, stream: Stream<T>): Stream<T> {
  return cb => {
    let count = 0
    return stream(x => {
      count++
      if (count > n) {
        cb(x)
      }
    })
  }
}

export function skipWhile<T>(
  predicate: (x: T) => boolean,
  stream: Stream<T>,
): Stream<T> {
  return cb => {
    let started = false
    return stream(x => {
      if (!started) {
        started = !predicate(x)
      }
      if (started) {
        cb(x)
      }
    })
  }
}

export function skipDuplicates<T>(
  comp: (prev: T, next: T) => boolean,
  stream: Stream<T>,
): Stream<T> {
  return sink => {
    let latest: Maybe<T> = {tag: "nothing"}
    return stream(x => {
      if (latest.tag === "nothing" || !comp(latest.value, x)) {
        latest = {tag: "just", value: x}
        sink(x)
      }
    })
  }
}

export function startWith<T, U>(x: T, stream: Stream<U>): Stream<T | U> {
  return cb => {
    cb(x)
    return stream(cb)
  }
}

export function multicast<T>(stream: Stream<T>): Stream<T> {
  let cbs: ((x: T) => void)[] = []
  const push = (x: T) => {
    cbs.forEach(sink => {
      if (cbs.indexOf(sink) !== -1) {
        sink(x)
      }
    })
  }
  let unsub: null | (() => void) = null
  return cb => {
    let disposed = false
    cbs = [...cbs, cb]
    if (cbs.length === 1) {
      unsub = stream(push)
    }
    return () => {
      if (disposed) {
        return
      }
      disposed = true
      const index = cbs.indexOf(cb)
      cbs = [...cbs.slice(0, index), ...cbs.slice(index + 1, cbs.length)]
      if (cbs.length === 0 && unsub !== null) {
        unsub()
        unsub = null
      }
    }
  }
}

/*

TODO:

later
interval
delay
throttle/debounce
sequentially (maybe add optional argument to fromIterable)

*/
