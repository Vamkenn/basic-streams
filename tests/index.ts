import {sum} from "../source/index"

it("example", () => {
  expect(sum(5, 1)).toMatchSnapshot()
})
