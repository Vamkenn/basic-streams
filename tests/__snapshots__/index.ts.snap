// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ap preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`ap preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`ap updates result when inputs update 1`] = `
Array [
  "after 10 ms...",
  11,
  "after 2 ms...",
  9,
  "after 8 ms...",
  4,
]
`;

exports[`chain preserves disposer of main sream 1`] = `
Array [
  Array [],
]
`;

exports[`chain preserves disposer of spawned streams 1`] = `
Array [
  Array [],
]
`;

exports[`chain result stream contains values from spawned at correct time 1`] = `
Array [
  "after 11 ms...",
  2,
  "after 1 ms...",
  3,
  "after 9 ms...",
  4,
  "after 1 ms...",
  6,
]
`;

exports[`chainLatest doesn't contain values emited from old spawned streams 1`] = `
Array [
  "after 16 ms...",
  2,
  "after 10 ms...",
  4,
  "after 6 ms...",
  6,
]
`;

exports[`chainLatest preserves disposer of main sream 1`] = `
Array [
  Array [],
]
`;

exports[`chainLatest preserves disposer of spawned streams 1`] = `
Array [
  Array [],
]
`;

exports[`chainLatest result stream contains values from spawned at correct time 1`] = `
Array [
  "after 11 ms...",
  2,
  "after 1 ms...",
  3,
  "after 9 ms...",
  4,
  "after 1 ms...",
  6,
]
`;

exports[`combineArray preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`combineArray preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`combineArray preserves disposers 3`] = `
Array [
  Array [],
]
`;

exports[`combineArray result stream contains values from sources 1`] = `
Array [
  "after 10 ms...",
  Array [
    1,
    3,
    5,
    7,
  ],
  "after 2 ms...",
  Array [
    1,
    4,
    5,
    7,
  ],
  "after 2 ms...",
  Array [
    2,
    4,
    5,
    7,
  ],
  "after 6 ms...",
  Array [
    2,
    4,
    6,
    7,
  ],
]
`;

exports[`combineArray works with one stream 1`] = `
Array [
  "after 7 ms...",
  Array [
    1,
  ],
  "after 7 ms...",
  Array [
    2,
  ],
]
`;

exports[`combineArray works with zero streams 1`] = `
Array [
  Array [],
]
`;

exports[`empty subscriber is never called 1`] = `Array []`;

exports[`filter preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`filter removes values 1`] = `
Array [
  "after 20 ms...",
  2,
]
`;

exports[`fromIterable works with array 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
]
`;

exports[`fromIterable works with generator 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
  Array [
    3,
  ],
]
`;

exports[`fromLoose callback is called 1`] = `
Array [
  "after 10 ms...",
  1,
  "after 10 ms...",
  2,
]
`;

exports[`fromLoose disposer is called 1`] = `
Array [
  Array [],
]
`;

exports[`fromLoose fixes stream #4: \`cb\` must be called with one argument 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`fromLoose fixes stream #6: After \`disposer\` was called, \`cb\` must not be called 1`] = `Array []`;

exports[`fromLoose fixes usage #1: \`stream\` must be called with one argument, \`cb\` 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`fromLoose fixes usage #4: \`disposer\` must be called with no arguments 1`] = `
Array [
  Array [],
]
`;

exports[`fromLoose fixes usage #5: \`disposer\` must be called at most once 1`] = `
Array [
  Array [],
]
`;

exports[`map modifies values with provided fn 1`] = `
Array [
  "after 10 ms...",
  2,
  "after 10 ms...",
  3,
]
`;

exports[`map preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`map2 preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`map2 preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`map2 updates result when inputs update 1`] = `
Array [
  "after 10 ms...",
  Array [
    1,
    3,
  ],
  "after 2 ms...",
  Array [
    2,
    3,
  ],
  "after 8 ms...",
  Array [
    2,
    4,
  ],
]
`;

exports[`map3 preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`map3 preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`map3 preserves disposers 3`] = `
Array [
  Array [],
]
`;

exports[`map3 updates result when inputs update 1`] = `
Array [
  "after 10 ms...",
  Array [
    1,
    3,
    5,
  ],
  "after 2 ms...",
  Array [
    1,
    4,
    5,
  ],
  "after 2 ms...",
  Array [
    2,
    4,
    5,
  ],
  "after 6 ms...",
  Array [
    2,
    4,
    6,
  ],
]
`;

exports[`merge preserves disposers 1`] = `
Array [
  Array [],
]
`;

exports[`merge preserves disposers 2`] = `
Array [
  Array [],
]
`;

exports[`merge preserves disposers 3`] = `
Array [
  Array [],
]
`;

exports[`merge result stream contains values from sources 1`] = `
Array [
  "after 6 ms...",
  3,
  "after 1 ms...",
  1,
  "after 3 ms...",
  5,
  "after 2 ms...",
  4,
  "after 2 ms...",
  2,
  "after 6 ms...",
  6,
]
`;

exports[`multicast first and second sunscribers get same async events 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`multicast first and second sunscribers get same async events 2`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`multicast first subscribers gets sync events 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`multicast preserves disposer (single subscriber) 1`] = `
Array [
  Array [],
]
`;

exports[`multicast preserves disposer (two subscribers) 1`] = `
Array [
  Array [],
]
`;

exports[`multicast second subscribers doesn't get sync events 1`] = `Array []`;

exports[`multicast subscriber doesn't get event after unsub() in response to that event 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`multicast unsub doesn't remove another version of same subscriber 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`of calls callback with the value 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`scan contains correct values 1`] = `
Array [
  Array [],
  "after 10 ms...",
  Array [
    1,
  ],
  "after 10 ms...",
  Array [
    1,
    2,
  ],
]
`;

exports[`scan preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`skip preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`skip returns equivalent stream if N is 0 1`] = `
Array [
  "after 10 ms...",
  1,
  "after 10 ms...",
  2,
  "after 10 ms...",
  3,
]
`;

exports[`skip skips first N items 1`] = `
Array [
  "after 30 ms...",
  3,
]
`;

exports[`skipDuplicates doesn't apply comparator to the first element 1`] = `Array []`;

exports[`skipDuplicates preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`skipDuplicates removes duplicates 1`] = `
Array [
  "after 10 ms...",
  1,
  "after 20 ms...",
  3,
]
`;

exports[`skipWhile preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`skipWhile returns empty stream if predicate is () => true 1`] = `Array []`;

exports[`skipWhile returns equivalent stream if predicate is () => false 1`] = `
Array [
  "after 10 ms...",
  1,
  "after 10 ms...",
  2,
  "after 10 ms...",
  3,
]
`;

exports[`skipWhile skips first items that satisfy predicate 1`] = `
Array [
  "after 30 ms...",
  3,
]
`;

exports[`startWith adds a value 1`] = `
Array [
  0,
  "after 10 ms...",
  1,
  "after 10 ms...",
  2,
]
`;

exports[`startWith preserves disposer 1`] = `
Array [
  Array [],
]
`;

exports[`take calls disposer of source stream when we dispose result stream 1`] = `
Array [
  Array [],
]
`;

exports[`take contains correct values at correct time 1`] = `
Array [
  "after 10 ms...",
  1,
]
`;

exports[`take doesn't call disposer twice 1`] = `
Array [
  Array [],
]
`;

exports[`take if n is 0 returns empty stream 1`] = `Array []`;

exports[`take subscribes to source even if n is 0 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`take takes first n and then calls disposer of source stream (async) 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`take takes first n and then calls disposer of source stream (sync) 1`] = `
Array [
  Array [],
]
`;

exports[`take takes first n and then calls disposer of source stream (sync) 2`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`takeUntil calls disposers after first value in controller 1`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers after first value in controller 2`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers properly (async value in controller) 1`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers properly (async value in controller) 2`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers when we dispose result stream erlier 1`] = `
Array [
  Array [],
]
`;

exports[`takeUntil calls disposers when we dispose result stream erlier 2`] = `
Array [
  Array [],
]
`;

exports[`takeUntil doesn't contain values after sync value from controller 1`] = `Array []`;

exports[`takeUntil doesn't contain values after value from controller 1`] = `
Array [
  "after 10 ms...",
  1,
  "after 10 ms...",
  2,
]
`;

exports[`takeUntil subscribes first to source and then to controller 1`] = `
Array [
  1,
  2,
]
`;

exports[`takeWhile calls disposer of source stream when we dispose result stream erlier 1`] = `
Array [
  Array [],
]
`;

exports[`takeWhile contains correct values at correct time 1`] = `
Array [
  "after 10 ms...",
  1,
  "after 10 ms...",
  2,
]
`;

exports[`takeWhile doesn't call disposer twice 1`] = `
Array [
  Array [],
]
`;

exports[`takeWhile return empty stream if predicate is () => false 1`] = `Array []`;

exports[`takeWhile takes values that satisfy predicate until first value that don't then calls disposer (async) 1`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;

exports[`takeWhile takes values that satisfy predicate until first value that don't then calls disposer (sync) 1`] = `
Array [
  Array [],
]
`;

exports[`takeWhile takes values that satisfy predicate until first value that don't then calls disposer (sync) 2`] = `
Array [
  Array [
    1,
  ],
  Array [
    2,
  ],
]
`;
